allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -23,6 +26,39 @@ def __iter__(self) -> Iterator[Instance]:
             raise ConfigurationError("For a lazy dataset reader, _read() must return a generator")
         return instances
 
+
+class _CachedLazyInstances(Iterable):
+    """
+    Like ``_LazyInstances``, but we take the actual input file that the instances come from, as
+    well as a cache file.  If the cache file exists, we read from that instead of from the main
+    file.  If the cache file `doesn't` exist, we create it on our first pass, and read it during
+    subsequent passes.
+    """
+    def __init__(self,
+                 read_from_file_method: Callable[[str], Iterable[Instance]],
+                 read_from_cache_method: Callable[[str], Iterable[Instance]],
+                 serialize_instance_method: Callable[[Instance], str],
+                 file_path: str,
+                 cache_file: str) -> None:
+        super().__init__()
+        self.read_from_file_method = read_from_file_method
+        self.read_from_cache_method = read_from_cache_method
+        self.serialize_instance_method = serialize_instance_method
+        self.file_path = file_path
+        self.cache_file = cache_file
+
+    def __iter__(self) -> Iterator[Instance]:
+        if os.path.exists(self.cache_file):
+            return self.read_from_cache_method(self.cache_file)
```

Comment:
I'm getting a mypy complaint on this line that I'm not sure about.  I think I maybe need to wrap this somehow, but I'm not certain.